### PR TITLE
Rehaul session setup

### PR DESCRIPTION
**Pull request by peterkxie**
NEED TO DISCUSS: How to handle local storage filling up
Option 1: Use sessionStorage over localStorage. sessionStorage persists across refresh and restore, but does not persist across different tabs and is cleared upon tab closing. Only when user hits save UI button does the current session get saved to localStorage. This way localstorage fills slower. Saved sessions refer to everything in local storage

Option 2: Use localstorage, add a saved key to the value saved in local storage. Currently, local storage looks like =
key: uuid
value: { sessionObject }
With this would be changed to
key: uuid
value: { session: {sessionObject}, saved: boolean }
Then on tab close, localStorage garbage collects itself and clears all objects without the saved: true. Saved get flipped to true when user hits save UI button. This options allows for the same current session to persist across tabs since localstorage has not run the garbage collection part yet

TODO: fix tests after above has been decided

This PR completely rehauls how sessions work. Instead of having a sessionQueryParam with the whole base64encoded session included in the URL, it works as follows
1. When user starts JBrowse, on new session it generates an uuid + current snapshot of the session and is saved to local storage. If the url contains a session param, it searches local storage for a matching uuid and loads that session
2. while user is working, the session is continually updated in local storage
3. when user wants to share, they click the share ui button which triggers a lambda function. The current session gets base64encoded and encrypted before being written to the DynamoDB. User will see a dialog box with their shareable URL, this URL contains the sessionID param that is used to look up the matching sessionId in DynamoDB, and it contains a password used to decrypt the data received back from the DB. This password only exists in the URL, no where else.
4. When another user receives the shared URL and loads the site, it will read from the DynamoDB using the sessionID, find the encrypted session, and send it back. Locally, the session will be decrypted and assigned a local uuid before being written into localStorage

savedSession has been changed, previously it was an array that was part of the jbrowseModel, and now it completely refers to the localStorage to find all savedSessions

default session naming has changed to be more readable, looks like New Session 2020-09-24 13:12:44

### CODE REVIEW

**Comment by peterkxie** [comment]
Notes from pairing:

Important to keep copying and pasting url into another tab and it loads
Possible paths to go from here -> use localstorage, only have a single obj in localstorage that is the most recent session that continually autosaves. any other session you want saved must have a generated share URL. savedSession concept is gone locally

use sessionStorage->everything is moved to sessionstorage, selecting save moves it to localstorage. localstorage also has 1 item where it is an autosave of the most recent session that has been changed. To send info between tabs, do POST calls so tabs can share their sessionStorage with each other (https://stackoverflow.com/questions/28230845/communication-between-tabs-or-windows#:~:text=Using%20postMessage%20to%20communicate%20between,to%20persist%20data%20being%20passed.&text=So%2C%20even%20if%20the%20opener,data%20even%20after%20getting%20refreshed)

**Comment by peterkxie** [comment]
If anyone wants to test out sessionStorage, the branch is 454_saved_sessions_sessionStorage

Quick concept of how it might work with usign sessionStorage and a save button

**Comment by cmdcolin** [comment]
@peterkxie this is great. I made an attempt to add a broadcast channel to share between tabs

this gives intuitive behavior of being able to copy your session to a new tab, but automatically creates a new session once in the new tab

**Comment by cmdcolin** [comment]
See https://github.com/GMOD/jbrowse-components/tree/add_broadcast_channel

**Comment by peterkxie** [comment]
added since last push: concept of using sessionStorage over localStorage

things to discuss: how autosave flow works, how overall flow works

File: packages/core/ui/NewSessionCards.js [hunk 1]
```diff
@@ -64,8 +64,12 @@ NewSessionCard.defaultProps = {
   style: {},
 }
 
+const tzOffset = new Date().getTimezoneOffset() * 60000 // tz in ms
+const dateString = `${new Date(Date.now() - tzOffset)
```
Comment (cmdcolin) [inline packages/core/ui/NewSessionCards.js:68]: Seems like new Date().toLocaleString() is approximately what we get from there?

**Comment by cmdcolin** [comment]
@peterkxie can you lay out the what is proposed by this PR in further detail, or at least in broad strokes

**Comment by cmdcolin** [comment]
I cloned this branch and I copied and pasted a URL from local into another tab and it gave me an error "No matching local session found" on a couple times and then sometimes it worked


Not sure if it was hitting the 1 second timeout for broadcast channel and I can't get it to happen again at this very moment but might be worth checking

**Comment by peterkxie** [comment]
Walkthrough of how the session changes are

BEFORE BEING SHARED OR SAVED:
On load, if there is no session param in the URL, one is generated in the format session=local-uuid and placed in session storage. If there were other objects in session storage, they are cleared out prior to this one being put in. This way, there can only be one item in session storage. Any changes you make on the session will record to this item in session storage. Also, any changes will write to an item in localstorage called autosave. The autosave object is overwritten on the most recent change, regardless of the current session the user is working on. The sessionStorage session is in an unsaved state, meaning that if the tab is closed, it is only recoverable by a Restore that is discussed later. When selecting 'New Session', a user will be prompted to save any unsaved changes before the previous session is wiped from sessionStorage and a new one with a new id is put in it's place.

ON SAVE:
A user will hit a save button if their sessionStorage session is ready for saving. It will then take the current session and write it into the localStorage with the key localSaved-uuid. The uuid is the same as the sessionStorage's uuid. This is because whenever a user hits save again on the same session, this localStorage object will be overwritten instead of creating another item in the localStorage. To manage these saved sessions, users can hit Open session and open or delete any of the saved sessions they may want, which will manage their localStorage for them graphically. On save, they will notice no changes in their URL since they are still working on their sessionStorage session, and will need to save again if they do big changes since their last save.

ON SHARE: 
When selecting share, the session information is first encrypted and a password is generated. This password is saved for the link generated later. The API gateway is then pinged with a POST request that contains the encrypted session, the current datetime when this request happens, and the referer. The gateway triggers a lambda that places this info with a sessionId into the DynamoDB table. The sessionId is just another hash of the session to give a unique identifier to find the session later. The DynamoDB only has a writeonce conditional, so no user can brute force write over another's session. Then, the lamdba returns the unique sessionId that points to the session that wants to be shared. the Url will contain something like this:
localhost/config=**&?session=share-**&?password=**. The session cannot be read without the password. If another user pastes this share link into the browser, it will call the API gateway with a GET request that tries to find a matching sessionId to the one in session= above. On success, the encrypted session is sent back and locally decrypted using the password in the URL. Then this session is placed in sessionStorage (cleared out before being placed) and is used like a normal local session.

AUTOSAVE AND RECOVERY:
As stated before, there is an autosave object that is created on any recent change. Once that exists, on any other load, if there is an existing autosave, the session in autosave will be written to localstorage in another object. This object serves as the recovery session. If user selects File -> Restore last autosave, it will load this object's session. This is because on any change autosave is overwritten, so the last autosave before it starts getting written by the current session is saved as a potential recovery point.

MISC:
If user copy's their URL and pastes to another tab before saving/sharing (the session is only located in sessionStorage), there will be a useEffect called that uses broadcast channels. The bc will look for tabs with a session storage that has the matching uuid and load that session into the new tab's sessionstorage

feel free to ask about other interactions, I probably missed a few

**Comment by cmdcolin** [comment]
I made several proposed modifications to this

See diff
https://github.com/GMOD/jbrowse-components/compare/454_saved_sessions_px...454_saved_sessions_colin


1) Gets rid of save in the header bar, I think that it is basically sort of a specialized function to have these saved sessions and so I move this into the File->Save session
2) Removes "Restore Autosave Session" and instead adds adds "Previous autosaved entry" to the Open session dialog
3) Makes it so that "Save session" saves to a local storage keyed on the session name, instead of "currentSessionId", which means that when you change your session name, and then save it, it is effectively a new session. This is sort of similar to maybe like "changing the filename", and is sort of an alternative to a "Save as" feature.
4) Adjusts onSnapshot in Loader.tsx to be onSnapshot(rootModel,...) instead of  onSnapshot(session,...). This is because a change to session.name was not recognized by onSnapshot(session, ...) and I wanted to snapshot logic to run on a session name change
5) Tries to avoid a proliferation of suffixes to the session name by removing -auto and adding -restored when we restore a previously autosaved session
6) Standardizes capitalization in the File menu as Capital first word, lower case other words. Also removes the "..." ellipses when it seems unneeded in Open session
7) Uses pluralize to make the logic in the SessionManager panel a little more straightforward. Reports only total tracks rather than tracks per panel, which is what the previous logic had
8) Makes the saved sessions an observable object so that the side panel updates in response to changes in the saved sessions

**Comment by cmdcolin** [comment]
Another alternative here is removing saved session functionality because it is a complex logic and if is not truly needed by this PR it may well be delivered better later

**Comment by garrettjstevens** [comment]
> Also removes the "..." ellipses when it seems unneeded in Open session

From [here](https://developer.apple.com/design/human-interface-guidelines/macos/menus/menu-anatomy/#menu-item-titles): 

> **Use an ellipsis whenever choosing a menu item requires additional input from the user.** The ellipsis character (…) means a dialog or separate window will open and prompt the user for additional information or to make a choice.

I think since "Open session" requires additional user input in the widget, it should have an ellipsis (similar to a standard "Open File" menu):

image

**Comment by rbuels** [comment]
Re: ellipsis: Google doesn't follow that convention, but nearly every other UI style does. I would say we should follow that convention.

**Comment by cmdcolin** [comment]
So...google is the material-ui gurus, so don't add it or do add it? I dunno I can go either way it's pretty minor. The other stuff is the more impactful changes

**Review by rbuels — COMMENTED** [review]
A few places where heavy lifting needs to be moved out of React files, and we need to make sure it doesn't die in Safari due to lack of BroadcastChannel support.

Otherwise, looks good! 👍

File: products/jbrowse-web/src/rootModel.ts [hunk 2]
```diff
@@ -44,52 +46,154 @@ export default function RootModel(
...
-        this.setSession({
+        const newSession = {
           ...self.jbrowse.defaultSession,
           name: `${self.jbrowse.defaultSession.name} ${new Date(
-            Date.now(),
-          ).toISOString()}`,
-        })
+            Date.now() - new Date().getTimezoneOffset() * 60000,
+          )
+            .toISOString()
```
[the remark this answers is not in the document]
Comment (cmdcolin) [reply products/jbrowse-web/src/rootModel.ts:84]: I suggested "new Date().toLocaleString()" previously in the thread

File: products/jbrowse-web/src/Loader.tsx [hunk 3]
```diff
@@ -105,14 +106,168 @@ function NoConfigMessage() {
...
+            // eslint-disable-next-line no-alert
+            alert('Session could not be decrypted with given password')
+            setData()
+          }
+        } else {
+          // eslint-disable-next-line no-alert
+          alert('Failed to find given session in database')
+          setData()
+        }
+      }
```
[the remark this answers is not in the document]
Comment (cmdcolin) [reply products/jbrowse-web/src/Loader.tsx:213]: So, if we think about this being someplace else, should it be a mobx-state-tree model? Should it just be utility functions that live outside the data function. It uses react hooks, session storage, fetches, lots of stuff going on

**Comment by rbuels** [comment]
@garrettjstevens @cmdcolin @peterkxie can we envision this session sharing being used in anything outside of jbrowse-web? should it be in core? This PR has much of it in core.

**Comment by peterkxie** [comment]
I put the encrypt/decrypt and read/write from dynamodb functions inside sessionSharing.js that lives in core for now, can move it out if needed

**Comment by rbuels** [comment]
Re: ellipsis, I think we should add ellipses.

**Comment by cmdcolin** [comment]
I have a couple more things to suggest for this. I will make a proposal shortly

**Comment by cmdcolin** [comment]
I made several posts to this branch that includes making the "localSaved" sessions use an observable. This tries to centralize the logic that is related to saving sessions to a single observable autorun. Even after this, deleting a session still needs to manually run localStorage.remove(item) because the observable can't autorun the delete

I also was concerned on this branch how scattered the use of sessionStorage and the manual setting of setSessionUuidInUrl

This seems like it should use our use-query-params and respond to a model state, rather than a manual action that we update

Also, the use of sessionStorage.clear() is probably a bad sign. A containing app could use sessionStorage and this would have a bad effect. We should aim to get rid of this

**Comment by cmdcolin** [comment]
I'm going to try a couple things to fix the aforementioned sessionStorage issue

**Comment by cmdcolin** [comment]
I made a couple of updates direct on this branch. I think it preserves the general functionality, can revert them if needed though.

If you want to retest that would be great @peterkxie  or @/all

This now uses shortid instead of uuid, stores an id direct on sessionModel as types.optional(types.identifier, shortid()) and it stores a single key in sessionStorage called "current" so it doesn't have to sessionStorage.clear()

I have tested the various aspects of this including shared session, saved session, restoring saved session, restoring autosave, and copying from tab to tab and it seemed to do ok.

**Comment by cmdcolin** [comment]
Garrett and I found a way to make it so the setQueryParam hook triggers a re-render and then refactored the Loader.tsx to be more resilient to having re-renders occur (currently it assumes that it will basically only ever run once after a config loaded to create the rootModel but accurately, it should be responsive to session param in the URL)

We made these changes here

**Comment by cmdcolin** [comment]
Forgot link https://github.com/GMOD/jbrowse-components/tree/454_saved_sessions_colin


Will probably do a little testing and merge into here

**Comment by peterkxie** [comment]
@cmdcolin I like the changes! I tested out all the workflows I had been doing previously (copy from tab, saving and deleting session, new session and recovering, sharing and loading shared session) and they work the same as before. Also like the change of moving everything out of core to jbrowse-web. Wasn't able to find the loader changes on the branch you linked but will check that out once it appears

**Comment by cmdcolin** [comment]
I actually pushed the loader changes to this branch now :)

**Comment by cmdcolin** [comment]
While this branch got to a working state, the loader file was getting quite complicated. It had 8 useEffects

I thought there could be an opportunity to take some of the logic out of there and put it into a mobx-state-tree type logic

While I won't say things are automatically better this way, it is in some ways easier to interpret than having multiple levels of useEffect that are interrelated by their dependency arrays (which is a particular badness when treating useEffect as kind of a pseudo-observable like this https://twitter.com/buildsghost/status/1281676305998557184?s=20)


Compare
https://github.com/GMOD/jbrowse-components/blob/454_saved_sessions_colin/products/jbrowse-web/src/Loader.tsx
https://github.com/GMOD/jbrowse-components/blob/454_saved_sessions_px/products/jbrowse-web/src/Loader.tsx

The logic is even still a little stringy but I think it is possibly a bit more organized, and maybe a bit more extensible since more complicated loading logic can now be added into this mst model instead of component rendering

**Comment by cmdcolin** [comment]
Merged the mst loader change onto this pr branch---i think it passed muster

**Comment by cmdcolin** [comment]
A change was added so that the URL's &config= variable is included in the saved sessions, autosave, and previousAutosave so that we don't get "cross contamination"

**Comment by cmdcolin** [comment]
I have restored the ability to generate and decode long URLs which are prefixed with &session=encoded-...

This allows an opportunity for programmatically generated sessions to still work, at least providing a fallback where a program would potentially generate these encoded URLs

**Comment by cmdcolin** [comment]
There is also a short "info" blurb on what the short URLs 



Note that  it may be "not as friendly" to be generating both and expect the user to understand what is going on there, but not doing this means we need to ensure total reliability of the URL shortener and global access which is not easy to do. I have also made it so the URL shortener could be set in the root config.json as

```{"configuration": {"shareURL": "http://baseofurlshortener/"},...}```
Where baseofurlshortener would be load and share functions

But to me, just generating both, allowing that this service could fail, is a good option
Screenshot from 2020-10-17 13-04-39


Screenshot from 2020-10-17 13-04-41

**Comment by rbuels** [comment]
couple of feedback items:
- if, down the road, the current encryption library we are using is shown to have a vulnerability, and we need to switch to a different one without breaking the existing cloud sessions, is there a straightforward path to implementing that?
- share dialog should have the "long" URL hidden by default, either in a collapsed section, or behind some kind of "mode" toggle switch, or tabs, or something similar

**Comment by cmdcolin** [comment]
>if, down the road, the current encryption library we are using is shown to have a vulnerability, and we need to switch to a different one without breaking the existing cloud sessions, is there a straightforward path to implementing that?

old sessions will not be able to be "reencoded", but we can switch to a newer version because the current response format is

{session:"ENCODEDCONTENT"} so a newer API could say {session:"ENCODEDCONTENT", encoding: "newencoding"}

>share dialog should have the "long" URL hidden by default, either in a collapsed section, or behind some kind of "mode" toggle switch, or tabs, or something similar

This is now a special settings dialog, and info about the sessions can be accessed in the settings dialog

localhost_3000__config=test_data%2Fvolvox%2Fconfig json session=auShUBrgH

localhost_3000__config=test_data%2Fvolvox%2Fconfig json session=auShUBrgH (1)
localhost_3000__config=test_data%2Fvolvox%2Fconfig json session=auShUBrgH (2)

**Comment by cmdcolin** [comment]
I marked this as ready now, I think it is "functionally complete"

There is some stuff related to session storage filling up and how to handle that which may need addressing

Review welcome

**Comment by cmdcolin** [comment]
Updated to have

a) local sessions prefixed with local-
b) a short blurb for example if someone shares you their local session with them
c) load default session on session loading error

localhost_3000__config=test_data%2Fvolvox%2Fconfig json session=local-RzHZF2JOT

**Comment by cmdcolin** [comment]
That above error is shown only if the session contained local-

If it is a short or long url that failed to load it will look like

localhost_3000__config=test_data%2Fvolvox%2Fconfig json session=local-Pbt7n_zpO

**Comment by cmdcolin** [comment]
This now prints out a very jbrowse 1 style screen when a config is not found

localhost_3000__config=doesNotExist json (2)


Also the list of sample data are pared down to just volvox and human data in production. I think it is useful even in production to provide our sample data links, as we do ship with them and they are good for confirming your instance works.

**Comment by cmdcolin** [comment]
I mentioned in my previous comment that volvox and human are shown in the production "NoConfigLoaded" screen, but I made it so it is only volvox since that is fully contained in a shipped instance folder. Human has links out to the web that could move or break over time

**Comment by cmdcolin** [comment]
This is going to now deploy to a new subfolder on s3

http://s3.amazonaws.com/jbrowse.org/code/jb2/alpha/master/index.html

changes to

http://s3.amazonaws.com/jbrowse.org/code/jb2/beta/master/index.html

Note that we will also obtain tagged releases like this once it gets going. The beta.26 is already there but had a bug that makes it fail to deploy to subfolders

http://s3.amazonaws.com/jbrowse.org/code/jb2/beta/@jbrowse%2Fweb@0.0.1-beta.27/index.html

**Comment by cmdcolin** [comment]
This is an added change that could be made
https://github.com/GMOD/jbrowse-components/compare/454_saved_sessions_px...454_saved_sessions_px2?expand=1


A) url-safe-base64 encode the password
B) emit a nano-id from the server side rather than a 256 byte hash from the server side. I see no particular reason to use hash of the session as the sessionId so nanoid helps



before
http://localhost:3000/?config=test_data%2Fvolvox%2Fconfig.json&session=share-0ce23826e79356b32bc00b01580765616eac25d349e60237df4a6d56c0f4f135&password=9b3a929aed770fdf56a5f8f7abe6fca6

after
http://localhost:3000/?config=test_data%2Fvolvox%2Fconfig.json&session=share-V1StGXR8_Z5jdHi6B-myT&password=ilpjn3iJ-9uRBu_RbSBFgw


that is a saving of 53 bytes so 183->130 characters

if interested, I can plug that change into the aws lambda and we can go from there

**Review by garrettjstevens — APPROVED** [review]
Looks good. I just noticed one small thing. If I do `File -> Save session` and then `File -> New session` without dismissing the "saved" snackbar, an empty warning snackbar pops up:

image

**Comment by cmdcolin** [comment]
awesome thanks for checking. I had run into that empty warning at least once and didn't know how to reproduce. the thing seems to be that our Snackbar component file holds onto a blank snackbar after the session changes

possible fix

```
diff --git a/packages/core/ui/Snackbar.tsx b/packages/core/ui/Snackbar.tsx
index 1bfdf5b39..77fbc2d95 100644
--- a/packages/core/ui/Snackbar.tsx
+++ b/packages/core/ui/Snackbar.tsx
@@ -19,7 +19,6 @@ function MessageSnackbar({
 }: {
   session: SnackbarSession & IAnyStateTreeNode
 }) {
-  const [open, setOpen] = useState(false)
   const [snackbarMessage, setSnackbarMessage] = useState<
     SnackbarMessage | undefined
   >()
@@ -37,15 +36,12 @@ function MessageSnackbar({
       if (!latestMessage) {
         setSnackbarMessage(undefined)
       } else if (snackbarMessage[0] !== latestMessage[0]) {
-        setOpen(false)
         timeoutId = setTimeout(() => {
           setSnackbarMessage(latestMessage)
-          setOpen(true)
         }, 100)
       }
     } else if (latestMessage) {
       setSnackbarMessage(latestMessage)
-      setOpen(true)
     }

     return () => {
@@ -58,13 +54,12 @@ function MessageSnackbar({
       return
     }
     popSnackbarMessage()
-    setOpen(false)
   }

   const [message, level] = snackbarMessage || []
   return (
     <Snackbar
-      open={open}
+      open={!!message}
       onClose={handleClose}
       action={

```

**Comment by garrettjstevens** [comment]
There's a little animation between consecutive snackbar messages that gets broken with that diff. How about this one:

```diff
diff --git a/packages/core/ui/Snackbar.tsx b/packages/core/ui/Snackbar.tsx
--- a/packages/core/ui/Snackbar.tsx
+++ b/packages/core/ui/Snackbar.tsx
@@ -64,7 +64,7 @@ function MessageSnackbar({
   const [message, level] = snackbarMessage || []
   return (
     <Snackbar
-      open={open}
+      open={open && !!message}
       onClose={handleClose}
       action={
         <IconButton aria-label="close" color="inherit" onClick={handleClose}>
```

**Comment by cmdcolin** [comment]
I went ahead and adopted the above suggestion so we get the animation

At the same time I removed an animation from ShareDialog

The entire ShareDialog was being rendered on every frame because of getSnapshot(session) inside an observer

**Comment by cmdcolin** [comment]
Alrighty, I will probably ahead with the merge. This has been a big adventure from both @peterxie and I and great input from the team. Hopefully it is a stable base to build off of.

@peterkxie did a big refactor to get all the pieces and unit testing of the loader in place and then we saw the Loader component  getting really overloaded so we did some refactoring to make the logic driven by the MST actions and observables, and we can add more arbitrary logic to this later. I could can imagine using this loader to do some advanced actions like programmatically generated URLs like &loc=GENEID or other stuff in the future

Great work all ;)

**Pull request merged**